## Introduction <a id='intro'></a>
In this project, I perform analysis on data from the entertainment industry. The dataset contains entries of movies and TV shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages 
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. Since there is no information about the quality of the data, preprocessing is completed before analysis.
 
The project consists of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

## Stage 1. Data overview <a id='data_review'></a>

Open and explore the data.

To make preprocessing and analysis easier, I will be using Pandas DataFrame.

In [1]:
import pandas as pd

I read the `movies_and_shows.csv` file from the `datasets` folder and save it in the `df` variable:

In [2]:
df = pd.read_csv('movies_and_shows.csv')

Printing the first 10 table rows:

In [3]:
df.head(10) 

name                      Character   r0le        TITLE   Type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

'info()' provides the general information about the table with one command.

In [4]:
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in this analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

There three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'. 


### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, I need to preprocess the data.

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
To preprocess the data, I correct the formatting in the column headers and deal with the missing values. Then, I check whether there are duplicates in the data.

In [5]:
df.columns 

Index(['   name', 'Character', 'r0le', 'TITLE', '  Type', 'release Year',
       'genres', 'imdb sc0re', 'imdb v0tes'],
      dtype='object')

Here I change the column names according to the rules of good style:
* If the name has several words, use snake_case
* All characters must be lowercase
* Remove whitespace
* Replace zero with letter 'o'

In [6]:
df = df.rename(columns = {
    '   name':'name',
    'Character':'character',
    'r0le':'role',
    'TITLE':'title',
    '  Type':'type',
    'release Year':'release_year',
    'imdb sc0re':'imdb_score',
    'imdb v0tes':'imdb_votes'
})

To check the result. I print the names of the columns once more:

In [7]:
df.columns

Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imdb_score', 'imdb_votes'],
      dtype='object')

### Missing values <a id='missing_values'></a>
First, I find the number of missing values in the table. To do so, I combine 'isna()' with 'sum()':

In [8]:
df.isna().sum() 

name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64

Not all missing values affect the research: the single missing value in `'title'` is not critical. The missing values in columns `'imdb_score'` and `'imdb_votes'` represent around 6% of all records (4,609 and 4,726, respectively, of the total 85,579). This could potentially affect our research. To avoid this issue, we will drop rows with missing values in the `'imdb_score'` and `'imdb_votes'` columns.

In [9]:
df.dropna(subset=['title','imdb_score','imdb_votes'],inplace=True)

To makee sure the table doesn't contain any more missing values, I count the missing values again.

In [10]:
df.isna().sum()

name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64

### Duplicates <a id='duplicates'></a>
Here I find the number of duplicate rows in the table.

In [11]:
df.duplicated().sum()

6994

I review the duplicate rows to determine if removing them would distort our dataset.

In [12]:
df[df.duplicated()].tail()

name                 character   role    title   type  \
85569       Jessica Cediel           Liliana Navarro  ACTOR  Lokillo  MOVIE   
85570  Javier Gardeaz?­bal  Agust??n "Peluca" Ort??z  ACTOR  Lokillo  MOVIE   
85571        Carla Giraldo            Valery Reinoso  ACTOR  Lokillo  MOVIE   
85572  Ana Mar??a S?­nchez                   Lourdes  ACTOR  Lokillo  MOVIE   
85577         Isabel Gaona                    Cacica  ACTOR  Lokillo  MOVIE   

       release_year      genres  imdb_score  imdb_votes  
85569          2021  ['comedy']         3.8        68.0  
85570          2021  ['comedy']         3.8        68.0  
85571          2021  ['comedy']         3.8        68.0  
85572          2021  ['comedy']         3.8        68.0  
85577          2021  ['comedy']         3.8        68.0

There are two clear duplicates in the printed rows. We can safely remove them.

In [13]:
df.drop_duplicates(inplace=True)


I check for duplicate rows once more to make sure I have removed all of them:

In [14]:
df.duplicated().sum()

0

Now I get rid of implicit duplicates in the `'type'` column. The string `'SHOW'` can be written in different ways. These kinds of errors will also affect the result.

Here I check the 'type' column for implicit duplicates. To do so I:
* Retrieve the intended dataframe column 
* Apply a sorting method to it
* For the sorted column, call the method that will return all unique column values

In [15]:
df['type'].sort_values().unique()

array(['MOVIE', 'SHOW', 'movies', 'shows', 'the movie', 'tv', 'tv series',
       'tv show', 'tv shows'], dtype=object)

Looking through the list, I find implicit duplicates of `'show'` (`'movie'` duplicates are ignored since the assumption is about shows).

I find the following implicit duplicates:
* `'shows'`
* `'SHOW'`
* `'tv show'`
* `'tv shows'`
* `'tv series'`
* `'tv'`

To get rid of them, I declare the function `replace_wrong_show()` with two parameters: 
* `wrong_shows_list=` — the list of duplicates
* `correct_show=` — the string with the correct value

The function corrects the names in the `'type'` column from the `df` table.

In [16]:
def replace_wrong_show(wrong_shows_list, correct_show):
    df['type'] = df['type'].replace(wrong_shows_list, correct_show)

I call `replace_wrong_show()` and pass it arguments so that it clears implicit duplicates and replaces them with `show`:

In [17]:
wrong_shows_list = ['shows', 'SHOW', 'tv show', 'tv shows', 'tv series', 'tv']
correct_show = 'show'
replace_wrong_show(wrong_shows_list, correct_show)

I make sure the duplicate names are removed. 

In [18]:
df['type'].unique()

array(['MOVIE', 'the movie', 'show', 'movies'], dtype=object)

### Conclusions <a id='data_preprocessing_conclusions'></a>
I detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed. 

The absence of duplicates will make the results more precise and easier to understand.

Now I can move on to my analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous project stages, I can now define how the assumption will be checked. I will calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

Based on this, I will complete the following steps:

- Filter the dataframe to only include shows released in 1999 or later.
- Group scores into buckets by rounding the values of the appropriate column (a set of 1-10 integers will help us make the outcome of our calculations more evident without damaging the quality of our research).
- Identify outliers among scores based on their number of votes, and exclude scores with few votes.
- Calculate the average votes for each score and check whether the assumption matches the results.

To filter the dataframe and only include shows released in 1999 or later, I will take two steps. First, I keep only titles published in 1999 or later in our dataframe. Then I filter the table to only contain shows (movies will be removed).

In [19]:
shows_1999_and_after = df[df['release_year'] >= 1999 ]

In [21]:
shows_1999_and_after = shows_1999_and_after[shows_1999_and_after['type'] == 'show']

The scores that are to be grouped should be rounded. For instance, titles with scores like 7.8, 8.1, and 8.3 will all be placed in the same bucket with a score of 8.

In [22]:
shows_1999_and_after['imdb_score'] = shows_1999_and_after['imdb_score'].round().astype(int)
shows_1999_and_after.tail()

name character      role      title  type  release_year  \
85433     Maneerat Kam-Uan        Ae     ACTOR  Let's Eat  show          2021   
85434    Rudklao Amratisha   unknown     ACTOR  Let's Eat  show          2021   
85435      Jaturong Mokjok   unknown     ACTOR  Let's Eat  show          2021   
85436     Pisamai Wilaisak   unknown     ACTOR  Let's Eat  show          2021   
85437  Sarawut Wichiensarn   unknown  DIRECTOR  Let's Eat  show          2021   

                    genres  imdb_score  imdb_votes  
85433  ['drama', 'comedy']           8         5.0  
85434  ['drama', 'comedy']           8         5.0  
85435  ['drama', 'comedy']           8         5.0  
85436  ['drama', 'comedy']           8         5.0  
85437  ['drama', 'comedy']           8         5.0

It is now time to identify outliers based on the number of votes.

In [23]:
shows_1999_and_after_grouped = shows_1999_and_after.groupby('imdb_score').nunique()
shows_1999_and_after_grouped

name  character  role  title  type  release_year  genres  \
imdb_score                                                             
2             24         18     2      5     1             4       5   
3             27         27     2      4     1             3       4   
4            174        108     2     30     1            13      19   
5            581        419     2     94     1            12      68   
6           2365       1896     2    347     1            19     212   
7           4342       3813     2    593     1            24     304   
8           4194       4028     2    539     1            23     329   
9            542        498     2     65     1            20      50   
10             8          8     1      1     1             1       1   

            imdb_votes  
imdb_score              
2                    5  
3                    4  
4                   30  
5                   91  
6                  329  
7                  568  
8                  527  
9                   65  
10                   1

Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores I use conditional filtering and grouping.

In [24]:
shows_1999_and_after = shows_1999_and_after[shows_1999_and_after['imdb_score'] >= 4]
shows_1999_and_after = shows_1999_and_after[shows_1999_and_after['imdb_score'] <= 9]

votes_mean = shows_1999_and_after.groupby('imdb_score')['imdb_votes'].mean().reset_index()
votes_mean

imdb_score     imdb_votes
0           4    5277.583333
1           5    3143.942568
2           6    3481.717322
3           7    8727.068211
4           8   30299.460967
5           9  126904.109515

Next I round the column with the averages, rename both columns, and print the dataframe in descending order.

In [25]:
# round column with averages
votes_mean['imdb_votes'] = votes_mean['imdb_votes'].round()

# rename columns
votes_mean.rename(columns={'imdb_votes':'average_votes'},inplace = True)

# print dataframe in descending order
votes_mean_sorted = votes_mean.sort_values(by='average_votes',ascending=False)
votes_mean_sorted

imdb_score  average_votes
5           9       126904.0
4           8        30299.0
3           7         8727.0
0           4         5278.0
2           6         3482.0
1           5         3144.0

The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.